test/download-dist

```diff
@@ -54,70 +54,40 @@ def download_dist(wait=False):
+        subprocess.check_call(["git", "clone", "--quiet", "--depth=1", "--bare", "https://github.com/" + REPO, dist_git_checkout])
+
     retries = 50 if wait else 1  # 25 minutes, once every 30s
     while retries > 0:
         try:
-            with urllib.request.urlopen(request) as response:
-                sys.stderr.write(f"download-dist: Downloading dist tarball from {download_url} ...\n")
-                if os.isatty(sys.stderr.fileno()):
-                    total_size = 0
-                else:
-                    total_size = None
-                MB = 10**6
-                # read tar into a stringio, as the stream is not seekable and tar requires that
-                while True:
-                    block = response.read(MB)
-                    if len(block) == 0:
-                        break
-                    if total_size is not None:
-                        total_size += len(block)
-                        sys.stderr.write(f"\r{ total_size // MB } MB")
-
-                    tario.write(block)
-
-                # clear the download progress in tty mode
-                if total_size is not None:
-                    sys.stderr.write("\r                             \r")
-
-                break
-
-        except urllib.error.HTTPError as e:
+            subprocess.check_call(["git", "-C", dist_git_checkout, "fetch", "origin", "sha-" + sha, "--depth=1"])
+            break
+        except subprocess.CalledProcessError:
             retries -= 1
 
             if retries == 0:
-                message(f"download-dist: Downloading {download_url} failed:", e)
+                message(f"download-dist: Downloading pre-built dist for SHA {sha} failed")
                 return False
 
-            message(f"download-dist: {download_url} not yet available, waiting...")
+            message(f"download-dist: pre-built dist for {sha} not yet available, waiting...")
             time.sleep(30)
 
-    tario.seek(0)
-    with tarfile.open(fileobj=tario) as ftar:
-        names = ftar.getnames()
-        try:
-            names.remove('.')
-        except ValueError:
-            pass
-        if len(names) != 1 or not names[0].endswith(".tar.gz"):
-            message("download-dist: expected tar with exactly one tar.gz member")
-            return False
-        ftar.extract(names[0])
-        tar_path = os.path.realpath(names[0])
-
-    # Extract relevant files
-    unpack_paths = ["dist", "package-lock.json"]
-    message("download-dist: Extracting paths from tarball:", ' '.join(unpack_paths))
-    prefixed_unpack_paths = ["cockpit-machines/" + d for d in unpack_paths]
-    subprocess.check_call(["tar", "--touch", "--strip-components=1", "-xf", tar_path] + prefixed_unpack_paths)
-
-    os.remove(tar_path)
+    message("download-dist: Extracting dist cache...")
+    p_git = subprocess.Popen(["git", "-C", dist_git_checkout, "archive", "FETCH_HEAD", "dist", "package-lock.json"],
```

Comment:
...or just use the shell?

Reply:
Well, maybe this is over-engineered indeed.. :grin: